Assets/Scripts/EyesightSystem/EyeSight.cs

```diff
@@ -0,0 +1,45 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEditor;
+using UnityEngine;
+using UnityEngine.Events;
+using UnityEngine.PlayerLoop;
+
+[Serializable]
+public class ObjectFoundEvent : UnityEvent<GameObject> {
+}
+
+public class EyeSight : MonoBehaviour {
+    public Vector2 sightDirection;
+    public Vector2 rayOffset;
+
+    [SerializeField] private float rayLength;
+    [SerializeField] private float sightAngle;
+    [SerializeField] private float sightAngleStep;
+    [SerializeField] private LayerMask sightMask;
+    [SerializeField] private LayerMask reportMask;
+    [SerializeField] private ObjectFoundEvent objectFoundEvent;
+
+    void FixedUpdate() {
+        float start = -sightAngle / 2;
+        float end = sightAngle / 2;
+
+        Vector2 startRaycast = new Vector2(transform.position.x + rayOffset.x,
+            transform.position.y + rayOffset.y);
+
+        for (; start <= end; start += sightAngleStep) {
+            Vector2 newDirection = new Vector2(sightDirection.x * Mathf.Cos(start * Mathf.Deg2Rad),
+                sightDirection.x * Mathf.Sin(start * Mathf.Deg2Rad));
+
+            Debug.DrawLine(startRaycast, startRaycast + newDirection * rayLength, Color.red);
+
+            RaycastHit2D hit = Physics2D.Raycast(startRaycast, newDirection, rayLength, sightMask);
+
+            if (hit.collider != null && ((1 << hit.collider.gameObject.layer) & reportMask) != 0) {
+                objectFoundEvent.Invoke(hit.collider.gameObject);
+                break;
+            }
```

Comment:
```suggestion
            if (!hit) {
                continue;
            }
            
            int objectLayerMask = 1 << hit.collider.gameObject.layer;
            if ((objectLayerMask & reportMask) == 0) {
                 continue;
            }
            
            objectFoundEvent.Invoke(hit.collider.gameObject);
            break;
```

`RayCastHit2D` implements a cast into `bool`, which is **cheaper** than `null` comparison. Reason for that is Unity object perverted stuff, which can be read [here](https://github.com/JetBrains/resharper-unity/wiki/Avoid-null-comparisons-against-UnityEngine.Object-subclasses) if you are curious... Yeah, that warning from Rider is serious.

I also think the if condition is looking too nested with all those parentheses, so I suggest splitting it as shown here.